Allow drop_last option in DistributedSampler

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#41171 Allow drop_last option in DistributedSampler**

Closes https://github.com/pytorch/pytorch/issues/25162. 

DistributedSampler allows data to be split evenly across workers in
DDP, but it has always added additional samples in order for the data to be
evenly split in the case that the # of samples is not evenly divisible by the
number of workers. This can cause issues such as when doing distributed
validation accuracy, where multiple samples could be considered twice. Also applications may not want to repeat the data within a single epoch when training and would rather drop the tail. 

This PR adds a drop_last option where the tail of the data is dropped such that
the effective dataset size is still evenly divisible across the workers. This
ensures that DDP can train fine (there is no uneven inputs) and each replica
gets an equal number of data indices.

The change is backwards compatible because the default value of `drop_last` is `False` so the old behavior is the default. 

Differential Revision: [D22449974](https://our.internmc.facebook.com/intern/diff/D22449974/)

rohan-varma committed:
Allow drop_last option in DistributedSampler

DistributedSampler allows data to be split evenly across workers in
DDP, but it has always added additional samples in order for the data to be
evenly split in the case that the # of samples is not evenly divisible by the
number of workers. This can cause issues such as when doing distributed
validation accuracy, where multiple samples could be considered twice.

This PR adds a drop_last option where the tail of the data is dropped such that
the effective dataset size is still evenly divisible across the workers. This
ensures that DDP can train fine (there is no uneven inputs) and each replica
gets an equal number of data indices.

Differential Revision: [D22449974](https://our.internmc.facebook.com/intern/diff/D22449974/)

[ghstack-poisoned]

mrshenli commented on torch/utils/data/distributed.py:
```diff
@@ -60,8 +64,18 @@ def __init__(self, dataset, num_replicas=None, rank=None, shuffle=True, seed=0):
+            # that each rank gets the same amount of data when iterating this
+            # dataloader.
+            self.num_samples = math.ceil((len(self.dataset) - self.num_replicas) / self.num_replicas)
+            self.total_size = self.num_samples * self.num_replicas
```
this line can be moved out of the if-else block?

mrshenli reviewed (commented):
LGTM! Test failure is real:

```
Jul 09 04:15:46 ======================================================================
Jul 09 04:15:46 ERROR [0.106s]: test_DistributedSampler_padding (__main__.TestDistBackend)
Jul 09 04:15:46 ----------------------------------------------------------------------
Jul 09 04:15:46 Traceback (most recent call last):
Jul 09 04:15:46   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 204, in wrapper
Jul 09 04:15:46     self._join_processes(fn)
Jul 09 04:15:46   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 311, in _join_processes
Jul 09 04:15:46     self._check_return_codes(elapsed_time)
Jul 09 04:15:46   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 344, in _check_return_codes
Jul 09 04:15:46     raise RuntimeError(error)
Jul 09 04:15:46 RuntimeError: Processes 0 exited with error code 10
```

zhaojuanmao commented on test/distributed/test_distributed.py:
```diff
@@ -2208,6 +2210,62 @@ def test_SyncBatchNorm_process_group(self):
         process_group_sync = res50_model_sync.layer1[0].bn1.process_group
         self.assertEqual(process_group_sync, process_group)
 
+    def test_DistributedSampler_padding(self):
```
nit: I'm wondering whether the test can be moved to data loader related test file?

rohan-varma replied:
I'd prefer to leave it in this fiile as `test_dataloader` doesn't have the multiprocessing setup, and we have the support for it in `test_distributed`, which we need since the test uses distributed collectives. Let me know if it's better to have this in the dataloader test though, and we can probably add a basic multiprocessing setup in the test to enable that.

zhaojuanmao reviewed (commented):
overall looks good to me, just one minor nit

rohan-varma committed:
Update on "Allow drop_last option in DistributedSampler"


Closes https://github.com/pytorch/pytorch/issues/25162. 

DistributedSampler allows data to be split evenly across workers in
DDP, but it has always added additional samples in order for the data to be
evenly split in the case that the # of samples is not evenly divisible by the
number of workers. This can cause issues such as when doing distributed
validation accuracy, where multiple samples could be considered twice. Also applications may not want to repeat the data within a single epoch when training and would rather drop the tail. 

This PR adds a drop_last option where the tail of the data is dropped such that
the effective dataset size is still evenly divisible across the workers. This
ensures that DDP can train fine (there is no uneven inputs) and each replica
gets an equal number of data indices.

The change is backwards compatible because the default value of `drop_last` is `False` so the old behavior is the default. 

Differential Revision: [D22449974](https://our.internmc.facebook.com/intern/diff/D22449974/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Allow drop_last option in DistributedSampler"


Closes https://github.com/pytorch/pytorch/issues/25162. 

DistributedSampler allows data to be split evenly across workers in
DDP, but it has always added additional samples in order for the data to be
evenly split in the case that the # of samples is not evenly divisible by the
number of workers. This can cause issues such as when doing distributed
validation accuracy, where multiple samples could be considered twice. Also applications may not want to repeat the data within a single epoch when training and would rather drop the tail. 

This PR adds a drop_last option where the tail of the data is dropped such that
the effective dataset size is still evenly divisible across the workers. This
ensures that DDP can train fine (there is no uneven inputs) and each replica
gets an equal number of data indices.

The change is backwards compatible because the default value of `drop_last` is `False` so the old behavior is the default. 

Differential Revision: [D22449974](https://our.internmc.facebook.com/intern/diff/D22449974/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Allow drop_last option in DistributedSampler"


Closes https://github.com/pytorch/pytorch/issues/25162. 

DistributedSampler allows data to be split evenly across workers in
DDP, but it has always added additional samples in order for the data to be
evenly split in the case that the # of samples is not evenly divisible by the
number of workers. This can cause issues such as when doing distributed
validation accuracy, where multiple samples could be considered twice. Also applications may not want to repeat the data within a single epoch when training and would rather drop the tail. 

This PR adds a drop_last option where the tail of the data is dropped such that
the effective dataset size is still evenly divisible across the workers. This
ensures that DDP can train fine (there is no uneven inputs) and each replica
gets an equal number of data indices.

The change is backwards compatible because the default value of `drop_last` is `False` so the old behavior is the default. 

Differential Revision: [D22449974](https://our.internmc.facebook.com/intern/diff/D22449974/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Allow drop_last option in DistributedSampler"


Closes https://github.com/pytorch/pytorch/issues/25162. 

DistributedSampler allows data to be split evenly across workers in
DDP, but it has always added additional samples in order for the data to be
evenly split in the case that the # of samples is not evenly divisible by the
number of workers. This can cause issues such as when doing distributed
validation accuracy, where multiple samples could be considered twice. Also applications may not want to repeat the data within a single epoch when training and would rather drop the tail. 

This PR adds a drop_last option where the tail of the data is dropped such that
the effective dataset size is still evenly divisible across the workers. This
ensures that DDP can train fine (there is no uneven inputs) and each replica
gets an equal number of data indices.

The change is backwards compatible because the default value of `drop_last` is `False` so the old behavior is the default. 

Differential Revision: [D22449974](https://our.internmc.facebook.com/intern/diff/D22449974/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Allow drop_last option in DistributedSampler"


Closes https://github.com/pytorch/pytorch/issues/25162. 

DistributedSampler allows data to be split evenly across workers in
DDP, but it has always added additional samples in order for the data to be
evenly split in the case that the # of samples is not evenly divisible by the
number of workers. This can cause issues such as when doing distributed
validation accuracy, where multiple samples could be considered twice. Also applications may not want to repeat the data within a single epoch when training and would rather drop the tail. 

This PR adds a drop_last option where the tail of the data is dropped such that
the effective dataset size is still evenly divisible across the workers. This
ensures that DDP can train fine (there is no uneven inputs) and each replica
gets an equal number of data indices.

The change is backwards compatible because the default value of `drop_last` is `False` so the old behavior is the default. 

Differential Revision: [D22449974](https://our.internmc.facebook.com/intern/diff/D22449974/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Allow drop_last option in DistributedSampler"


Closes https://github.com/pytorch/pytorch/issues/25162. 

DistributedSampler allows data to be split evenly across workers in
DDP, but it has always added additional samples in order for the data to be
evenly split in the case that the # of samples is not evenly divisible by the
number of workers. This can cause issues such as when doing distributed
validation accuracy, where multiple samples could be considered twice. Also applications may not want to repeat the data within a single epoch when training and would rather drop the tail. 

This PR adds a drop_last option where the tail of the data is dropped such that
the effective dataset size is still evenly divisible across the workers. This
ensures that DDP can train fine (there is no uneven inputs) and each replica
gets an equal number of data indices.

The change is backwards compatible because the default value of `drop_last` is `False` so the old behavior is the default. 

Differential Revision: [D22449974](https://our.internmc.facebook.com/intern/diff/D22449974/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Allow drop_last option in DistributedSampler"


Closes https://github.com/pytorch/pytorch/issues/25162. 

DistributedSampler allows data to be split evenly across workers in
DDP, but it has always added additional samples in order for the data to be
evenly split in the case that the # of samples is not evenly divisible by the
number of workers. This can cause issues such as when doing distributed
validation accuracy, where multiple samples could be considered twice. Also applications may not want to repeat the data within a single epoch when training and would rather drop the tail. 

This PR adds a drop_last option where the tail of the data is dropped such that
the effective dataset size is still evenly divisible across the workers. This
ensures that DDP can train fine (there is no uneven inputs) and each replica
gets an equal number of data indices.

The change is backwards compatible because the default value of `drop_last` is `False` so the old behavior is the default. 

Differential Revision: [D22449974](https://our.internmc.facebook.com/intern/diff/D22449974/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Allow drop_last option in DistributedSampler"


Closes https://github.com/pytorch/pytorch/issues/25162. 

DistributedSampler allows data to be split evenly across workers in
DDP, but it has always added additional samples in order for the data to be
evenly split in the case that the # of samples is not evenly divisible by the
number of workers. This can cause issues such as when doing distributed
validation accuracy, where multiple samples could be considered twice. Also applications may not want to repeat the data within a single epoch when training and would rather drop the tail. 

This PR adds a drop_last option where the tail of the data is dropped such that
the effective dataset size is still evenly divisible across the workers. This
ensures that DDP can train fine (there is no uneven inputs) and each replica
gets an equal number of data indices.

The change is backwards compatible because the default value of `drop_last` is `False` so the old behavior is the default. 

Differential Revision: [D22449974](https://our.internmc.facebook.com/intern/diff/D22449974/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Allow drop_last option in DistributedSampler"


Closes https://github.com/pytorch/pytorch/issues/25162. 

DistributedSampler allows data to be split evenly across workers in
DDP, but it has always added additional samples in order for the data to be
evenly split in the case that the # of samples is not evenly divisible by the
number of workers. This can cause issues such as when doing distributed
validation accuracy, where multiple samples could be considered twice. Also applications may not want to repeat the data within a single epoch when training and would rather drop the tail. 

This PR adds a drop_last option where the tail of the data is dropped such that
the effective dataset size is still evenly divisible across the workers. This
ensures that DDP can train fine (there is no uneven inputs) and each replica
gets an equal number of data indices.

The change is backwards compatible because the default value of `drop_last` is `False` so the old behavior is the default. 

Differential Revision: [D22449974](https://our.internmc.facebook.com/intern/diff/D22449974/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Allow drop_last option in DistributedSampler"


Closes https://github.com/pytorch/pytorch/issues/25162. 

DistributedSampler allows data to be split evenly across workers in
DDP, but it has always added additional samples in order for the data to be
evenly split in the case that the # of samples is not evenly divisible by the
number of workers. This can cause issues such as when doing distributed
validation accuracy, where multiple samples could be considered twice. Also applications may not want to repeat the data within a single epoch when training and would rather drop the tail. 

This PR adds a drop_last option where the tail of the data is dropped such that
the effective dataset size is still evenly divisible across the workers. This
ensures that DDP can train fine (there is no uneven inputs) and each replica
gets an equal number of data indices.

The change is backwards compatible because the default value of `drop_last` is `False` so the old behavior is the default. 

Differential Revision: [D22449974](https://our.internmc.facebook.com/intern/diff/D22449974/)

[ghstack-poisoned]
